File: packages/client/src/first-contact.ts
import {
  FIRST_CONTACT_PROTOCOL_VERSION,
  type FirstContactBinding
} from "@mdbase-dev/connect-protocol";

const TRANSCRIPT_DOMAIN = new TextEncoder().encode(
  "mdbase-connect first-contact transcript\0"
);
const SAS_INFO = new TextEncoder().encode(
  "mdbase-connect first-contact sas v1"
);
const CROCKFORD_BASE32 = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
const UUID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export type FirstContactRole = "application" | "connector";

export class FirstContactCryptoError extends Error {
  constructor(public readonly code: string, message: string) {
    super(message);
  }
}

/**
 * Derive the independently displayed first-contact authentication string.
 * The private key must belong to the endpoint role named in the transcript.
 */
export async function deriveFirstContactSas(
  binding: FirstContactBinding,
  role: FirstContactRole,
  identity: {
    agreementPrivateKey: CryptoKey;
    agreementPublicKey: string;
  }
): Promise<string> {
  const keys = await validatedKeys(binding);
  const own = role === "application"
    ? binding.application_agreement_public_key
    : binding.connector_agreement_public_key;
  const peer = role === "application"
    ? keys.connector
    : keys.applicationAgreement;
  if (identity.agreementPublicKey !== own) {
    throw new FirstContactCryptoError(
      "identity_mismatch",
      "The first-contact identity does not match the transcript."
    );
  }
  let peerKey: CryptoKey;
  try {
    peerKey = await crypto.subtle.importKey(
      "raw",
      peer as BufferSource,
      { name: "ECDH", namedCurve: "P-256" },
      false,
      []
    );
  } catch {
    throw new FirstContactCryptoError(
      "invalid_public_key",
      "The first-contact public key is invalid."
    );
  }
  const shared = await crypto.subtle.deriveBits(
    { name: "ECDH", public: peerKey },
    identity.agreementPrivateKey,
    256
  );
  const transcript = firstContactTranscript(binding, keys);
  const salt = await crypto.subtle.digest("SHA-256", transcript as BufferSource);
  const hkdfKey = await crypto.subtle.importKey(
    "raw",
    shared,
    "HKDF",
    false,
    ["deriveBits"]
  );
  const sas = new Uint8Array(await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt, info: SAS_INFO },
    hkdfKey,
    40
  ));
  return formatSas(sas);
}

interface ValidatedKeys {
  applicationAgreement: Uint8Array;
  applicationSigning: Uint8Array;
  connector: Uint8Array;
}

async function validatedKeys(
  binding: FirstContactBinding
): Promise<ValidatedKeys> {
  if (binding.protocol_version !== FIRST_CONTACT_PROTOCOL_VERSION) {
    throw new FirstContactCryptoError(
      "unsupported_version",
      "The first-contact protocol version is not supported."
    );
  }
  const applicationAgreement = canonicalPublicKey(
    binding.application_agreement_public_key
  );
  const applicationSigning = canonicalPublicKey(
    binding.application_signing_public_key
  );
  const connector = canonicalPublicKey(
    binding.connector_agreement_public_key
  );
  if (
    equalBytes(applicationAgreement, applicationSigning)
    || equalBytes(applicationAgreement, connector)
    || equalBytes(applicationSigning, connector)
  ) {
    throw new FirstContactCryptoError(
      "invalid_public_key",
      "First contact requires distinct endpoint agreement and signing keys."
    );
  }
  try {
    await Promise.all([
      applicationAgreement,
      applicationSigning,
      connector
    ].map((key) => crypto.subtle.importKey(
      "raw",
      key as BufferSource,
      { name: "ECDH", namedCurve: "P-256" },
      false,
      []
    )));
  } catch {
    throw new FirstContactCryptoError(
      "invalid_public_key",
      "The first-contact public key is invalid."
    );
  }
  return { applicationAgreement, applicationSigning, connector };
}

function firstContactTranscript(
  binding: FirstContactBinding,
  keys: ValidatedKeys
): Uint8Array {
  const chunks = [
    TRANSCRIPT_DOMAIN,
    u32(binding.protocol_version),
    field(uuidBytes(binding.application_id)),
    field(uuidBytes(binding.application_installation_id)),
    field(keys.applicationAgreement),
    field(keys.applicationSigning),
    field(uuidBytes(binding.connector_id)),
    field(keys.connector)
  ];
  const length = chunks.reduce((total, chunk) => total + chunk.byteLength, 0);
  const output = new Uint8Array(length);
  let offset = 0;
  for (const chunk of chunks) {
    output.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return output;
}

function field(value: Uint8Array): Uint8Array {
  const output = new Uint8Array(4 + value.byteLength);
  output.set(u32(value.byteLength));
  output.set(value, 4);
  return output;
}

function u32(value: number): Uint8Array {
  const output = new Uint8Array(4);
  new DataView(output.buffer).setUint32(0, value, false);
  return output;
}

function uuidBytes(value: string): Uint8Array {
  if (!UUID_PATTERN.test(value)) {
    throw new FirstContactCryptoError(
      "invalid_binding",
      "The first-contact transcript contains an invalid identifier."
    );
  }
  const compact = value.replaceAll("-", "");
  return Uint8Array.from(
    { length: 16 },
    (_, index) => Number.parseInt(compact.slice(index * 2, index * 2 + 2), 16)
  );
}

function canonicalPublicKey(value: string): Uint8Array {
  if (!/^[A-Za-z0-9_-]+$/.test(value)) {
    throw new FirstContactCryptoError(
      "invalid_public_key",
      "The first-contact public key is invalid."
    );
  }
  const normalized = value.replaceAll("-", "+").replaceAll("_", "/");
  const padded = normalized.padEnd(Math.ceil(normalized.length / 4) * 4, "=");
  let bytes: Uint8Array;
  try {
    bytes = Uint8Array.from(atob(padded), (character) => character.charCodeAt(0));
  } catch {
    throw new FirstContactCryptoError(
      "invalid_public_key",
      "The first-contact public key is invalid."
    );
  }
  if (
    bytes.byteLength !== 65
    || bytes[0] !== 4
    || base64Url(bytes) !== value
  ) {
    throw new FirstContactCryptoError(
      "invalid_public_key",
      "The first-contact public key is invalid."
    );
  }
  return bytes;
}

function base64Url(value: Uint8Array): string {
  return btoa(String.fromCharCode(...value))
    .replaceAll("+", "-")
    .replaceAll("/", "_")
    .replace(/=+$/u, "");
}

function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  return left.byteLength === right.byteLength
    && left.every((value, index) => value === right[index]);
}

function formatSas(value: Uint8Array): string {
  let bits = 0n;
  for (const byte of value) bits = (bits << 8n) | BigInt(byte);
  let output = "";
  for (let index = 7; index >= 0; index -= 1) {
    output += CROCKFORD_BASE32[Number((bits >> BigInt(index * 5)) & 31n)];
    if (index === 4) output += "-";
  }
  return output;
}

File: crates/connect-protocol/src/first_contact.rs
use crate::crypto::RelayIdentity;
use crate::FIRST_CONTACT_PROTOCOL_VERSION;
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use hkdf::Hkdf;
use p256::PublicKey;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use thiserror::Error;
use uuid::Uuid;

const TRANSCRIPT_DOMAIN: &[u8] = b"mdbase-connect first-contact transcript\0";
const SAS_INFO: &[u8] = b"mdbase-connect first-contact sas v1";
const CROCKFORD_BASE32: &[u8; 32] = b"0123456789ABCDEFGHJKMNPQRSTVWXYZ";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FirstContactBinding {
    pub protocol_version: u32,
    pub application_id: Uuid,
    pub application_installation_id: Uuid,
    pub application_agreement_public_key: String,
    pub application_signing_public_key: String,
    pub connector_id: Uuid,
    pub connector_agreement_public_key: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FirstContactRole {
    Application,
    Connector,
}

#[derive(Debug, Error, PartialEq, Eq)]
pub enum FirstContactError {
    #[error("unsupported first-contact protocol version")]
    UnsupportedVersion,
    #[error("invalid first-contact public key")]
    InvalidPublicKey,
    #[error("first-contact identity does not match the transcript")]
    IdentityMismatch,
    #[error("first-contact key derivation failed")]
    DerivationFailed,
}

impl FirstContactBinding {
    pub fn validate(&self) -> Result<(), FirstContactError> {
        self.validated_keys().map(|_| ())
    }

    pub fn derive_sas(
        &self,
        identity: &RelayIdentity,
        role: FirstContactRole,
    ) -> Result<String, FirstContactError> {
        let keys = self.validated_keys()?;
        let (own, peer) = match role {
            FirstContactRole::Application => (&keys.application_agreement, &keys.connector),
            FirstContactRole::Connector => (&keys.connector, &keys.application_agreement),
        };
        if identity.public_key() != URL_SAFE_NO_PAD.encode(own) {
            return Err(FirstContactError::IdentityMismatch);
        }
        let shared = identity
            .shared_secret(&URL_SAFE_NO_PAD.encode(peer))
            .map_err(|_| FirstContactError::InvalidPublicKey)?;
        let transcript_hash = Sha256::digest(self.transcript(&keys));
        let hkdf = Hkdf::<Sha256>::new(Some(&transcript_hash), &shared);
        let mut sas = [0_u8; 5];
        hkdf.expand(SAS_INFO, &mut sas)
            .map_err(|_| FirstContactError::DerivationFailed)?;
        Ok(format_sas(sas))
    }

    fn validated_keys(&self) -> Result<ValidatedKeys, FirstContactError> {
        if self.protocol_version != FIRST_CONTACT_PROTOCOL_VERSION {
            return Err(FirstContactError::UnsupportedVersion);
        }
        let application_agreement = canonical_public_key(&self.application_agreement_public_key)?;
        let application_signing = canonical_public_key(&self.application_signing_public_key)?;
        let connector = canonical_public_key(&self.connector_agreement_public_key)?;
        if application_agreement == application_signing
            || application_agreement == connector
            || application_signing == connector
        {
            return Err(FirstContactError::InvalidPublicKey);
        }
        Ok(ValidatedKeys {
            application_agreement,
            application_signing,
            connector,
        })
    }

    fn transcript(&self, keys: &ValidatedKeys) -> Vec<u8> {
        let mut transcript = Vec::with_capacity(300);
        transcript.extend_from_slice(TRANSCRIPT_DOMAIN);
        transcript.extend_from_slice(&self.protocol_version.to_be_bytes());
        append_field(&mut transcript, self.application_id.as_bytes());
        append_field(&mut transcript, self.application_installation_id.as_bytes());
        append_field(&mut transcript, &keys.application_agreement);
        append_field(&mut transcript, &keys.application_signing);
        append_field(&mut transcript, self.connector_id.as_bytes());
        append_field(&mut transcript, &keys.connector);
        transcript
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApplicationTrustPresentation {
    pub application_name: String,
    pub application_distribution: String,
    pub application_homepage: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub application_project_url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub application_icon: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApplicationTrustRequest {
    pub request_id: Uuid,
    pub binding: FirstContactBinding,
    pub presentation: ApplicationTrustPresentation,
    pub created_at: String,
    pub expires_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApplicationTrust {
    pub id: Uuid,
    pub binding: FirstContactBinding,
    pub presentation: ApplicationTrustPresentation,
    pub trusted_at: String,
    pub last_used_at: String,
}

struct ValidatedKeys {
    application_agreement: Vec<u8>,
    application_signing: Vec<u8>,
    connector: Vec<u8>,
}

fn canonical_public_key(encoded: &str) -> Result<Vec<u8>, FirstContactError> {
    let decoded = URL_SAFE_NO_PAD
        .decode(encoded)
        .map_err(|_| FirstContactError::InvalidPublicKey)?;
    if URL_SAFE_NO_PAD.encode(&decoded) != encoded
        || decoded.len() != 65
        || decoded.first() != Some(&4)
        || PublicKey::from_sec1_bytes(&decoded).is_err()
    {
        return Err(FirstContactError::InvalidPublicKey);
    }
    Ok(decoded)
}

fn append_field(output: &mut Vec<u8>, field: &[u8]) {
    output.extend_from_slice(&(field.len() as u32).to_be_bytes());
    output.extend_from_slice(field);
}

fn format_sas(bytes: [u8; 5]) -> String {
    let mut encoded = String::with_capacity(9);
    let value = u64::from_be_bytes([0, 0, 0, bytes[0], bytes[1], bytes[2], bytes[3], bytes[4]]);
    for index in (0..8).rev() {
        let digit = ((value >> (index * 5)) & 31) as usize;
        encoded.push(CROCKFORD_BASE32[digit] as char);
        if index == 4 {
            encoded.push('-');
        }
    }
    encoded
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fixture() -> (FirstContactBinding, RelayIdentity, RelayIdentity, String) {
        let fixture: serde_json::Value = serde_json::from_str(include_str!(
            "../../../packages/protocol/test/fixtures/first-contact-v1.json"
        ))
        .unwrap();
        let binding = serde_json::from_value(fixture["binding"].clone()).unwrap();
        let application = RelayIdentity::from_bytes(
            &URL_SAFE_NO_PAD
                .decode(fixture["application_private_key"].as_str().unwrap())
                .unwrap(),
        )
        .unwrap();
        let connector = RelayIdentity::from_bytes(
            &URL_SAFE_NO_PAD
                .decode(fixture["connector_private_key"].as_str().unwrap())
                .unwrap(),
        )
        .unwrap();
        (
            binding,
            application,
            connector,
            fixture["sas"].as_str().unwrap().to_string(),
        )
    }

    fn scalar(value: u8) -> [u8; 32] {
        let mut bytes = [0_u8; 32];
        bytes[31] = value;
        bytes
    }

    #[test]
    fn endpoints_derive_the_same_sas() {
        let (binding, application, connector, expected) = fixture();
        let application_sas = binding
            .derive_sas(&application, FirstContactRole::Application)
            .unwrap();
        let connector_sas = binding
            .derive_sas(&connector, FirstContactRole::Connector)
            .unwrap();
        assert_eq!(application_sas, connector_sas);
        assert_eq!(application_sas, expected);
    }

    #[test]
    fn substituted_keys_change_the_sas() {
        let (binding, application, connector, _) = fixture();
        let honest = binding
            .derive_sas(&application, FirstContactRole::Application)
            .unwrap();
        let attacker = RelayIdentity::from_bytes(&scalar(4)).unwrap();
        let mut substituted = binding;
        substituted.connector_agreement_public_key = attacker.public_key();
        let attacked = substituted
            .derive_sas(&application, FirstContactRole::Application)
            .unwrap();
        assert_ne!(honest, attacked);
        assert_eq!(
            substituted.derive_sas(&connector, FirstContactRole::Connector),
            Err(FirstContactError::IdentityMismatch)
        );
    }

    #[test]
    fn rejects_noncanonical_and_reused_keys() {
        let (mut binding, application, _, _) = fixture();
        binding.application_signing_public_key = binding.application_agreement_public_key.clone();
        assert_eq!(
            binding.derive_sas(&application, FirstContactRole::Application),
            Err(FirstContactError::InvalidPublicKey)
        );

        binding.application_signing_public_key = "not-a-key".to_string();
        assert_eq!(
            binding.derive_sas(&application, FirstContactRole::Application),
            Err(FirstContactError::InvalidPublicKey)
        );
    }
}
